fri/src/layers.rs

```diff
@@ -0,0 +1,378 @@
+    evaluation: Vec<F>,
+}
+
+#[allow(dead_code)]
+impl<F: PrimeField, E: EvaluationDomain<F>> FriLayerReal<F, E> {
+    pub fn new(domain: E, evaluation: Vec<F>) -> Self {
+        Self {
+            layer_size: domain.size(),
+            domain,
+            evaluation,
+        }
+    }
+
+    pub fn new_from_prev_layer(prev_layer: &dyn FriLayer<F, E>) -> Self {
+        Self {
+            layer_size: prev_layer.get_layer_size(),
+            domain: prev_layer.get_domain(),
+            evaluation: prev_layer.get_layer().unwrap(),
+        }
+    }
+}
+
+impl<F: PrimeField, E: EvaluationDomain<F> + 'static> FriLayer<F, E> for FriLayerReal<F, E> {
+    fn get_layer_size(&self) -> usize {
+        self.layer_size
+    }
+
+    fn get_domain(&self) -> E {
+        self.domain
+    }
+
+    fn get_layer(&self) -> Result<Vec<F>, Error> {
+        Ok(self.evaluation.clone())
+    }
+
+    fn eval_at_points(&self, required_indices: &[usize]) -> Vec<F> {
+        required_indices
+            .iter()
+            .map(|&i| self.evaluation[i])
+            .collect()
+    }
+
+    fn clone_arc(&self) -> Arc<dyn FriLayer<F, E>> {
+        Arc::new(self.clone())
+    }
+}
+
+pub struct FriLayerProxy<F: PrimeField, E: EvaluationDomain<F>> {
+    layer_size: usize,
+    domain: E,
+    prev_layer: Arc<dyn FriLayer<F, E>>,
+    eval_point: F,
+}
+
+#[allow(dead_code)]
+impl<F: PrimeField, E: EvaluationDomain<F>> FriLayerProxy<F, E> {
+    pub fn new(prev_layer: Arc<dyn FriLayer<F, E>>, eval_point: F) -> Self {
+        let prev_layer_domain = prev_layer.get_domain();
+        let current_domain = {
+            assert!(prev_layer_domain.size().is_power_of_two());
+            let coset_size = prev_layer_domain.size() / 2;
+            let coset_offset = prev_layer_domain.coset_offset().square();
+            E::new_coset(coset_size, coset_offset).unwrap()
+        };
+
+        Self {
+            layer_size: current_domain.size() as usize,
+            domain: current_domain,
+            prev_layer,
+            eval_point,
+        }
+    }
+}
+
+impl<F: PrimeField, E: EvaluationDomain<F> + 'static> FriLayer<F, E> for FriLayerProxy<F, E> {
+    fn get_layer_size(&self) -> usize {
+        self.layer_size
+    }
+
+    fn get_domain(&self) -> E {
+        self.domain
+    }
+
+    fn get_layer(&self) -> Result<Vec<F>, Error> {
+        let prev_layer_domain = self.prev_layer.get_domain();
+        let prev_eval = self.prev_layer.get_layer()?;
+
+        MultiplicativeFriFolder::compute_next_fri_layer(
+            prev_layer_domain,
+            &prev_eval,
+            &self.eval_point,
+        )
+    }
+
+    fn eval_at_points(&self, _required_indices: &[usize]) -> Vec<F> {
+        unimplemented!("Should never be called")
+    }
+
+    fn clone_arc(&self) -> Arc<dyn FriLayer<F, E>> {
+        Arc::new(FriLayerProxy {
+            layer_size: self.layer_size,
+            domain: self.domain,
+            prev_layer: Arc::clone(&self.prev_layer),
```

Comment:
`self.prev_layer.clone()` does the same.

Reply:
Removed the `clone_arc` function.